File: LeetCode/234. Palindrome Linked List/Solution.cpp
/**
 * Given a singly linked list, determine if it is a palindrome.
 * Example 1:
 * 
 * Input: 1->2
 * Output: false
 * Example 2:
 * 
 * Input: 1->2->2->1
 * Output: true
 */

#include <iostream>
#include <stack>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if (head == NULL || head->next == NULL)
            return true;
        
        stack<int> data;
        
        ListNode* slow = head;
        ListNode* fast = head;
        
        while(fast != NULL && fast->next != NULL)
        {
            data.push(slow->val);
            slow = slow->next;
            fast = fast->next->next;
        }
        
        if (fast != NULL)
            slow = slow->next;
        
        while(slow != NULL)
        {
            if (data.top() != slow->val)
                return false;
            
            data.pop();
            slow = slow->next;
        }
        
        return true;
    }
};

File: LeetCode/234. Palindrome Linked List/Solution.java
/**
 * Given a singly linked list, determine if it is a palindrome.
 * Example 1:
 * 
 * Input: 1->2
 * Output: false
 * Example 2:
 * 
 * Input: 1->2->2->1
 * Output: true
 */

 /**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null)
            return true;
            
        Stack<Integer> data = new Stack<Integer>();
        
        ListNode slow = head;
        ListNode fast = head;
        
        while (fast != null && fast.next != null)
        {
            data.push(slow.val);
            slow = slow.next;
            fast = fast.next.next;
        }
        
        // handles the case when there is an odd number of elements.
        // The middle element is required to be alone, so no need to pump it into the stack.
        // fast will not be null for odd number of elements since we are incrementing it by 2.
        
        if (fast != null)
            slow = slow.next;
        
        while(slow != null)
        {
            if(data.pop() != slow.val)
                return false;
            
            slow = slow.next;
        }
        
        return true;
        
    }
}